### ISSUE TITLE
[Patch] Bad boundary check when parsing a Socks4a packet

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2791]
|<img alt="adam.lofts's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| adam.lofts reported|
|-|-|
|Trac ID|trac#2791|
|Type|enhancement|
|Created|2007-08-19 16:17:05Z|

The current socks parsing code checks for one null after the header. However if the packet is a socks 4a packet the code assumes another null after the hostname. Thus the code breaks if the buffer does not yet contain a whole packet. 

Attachments:

* [socks_bad_bounday_check.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/66d/66d5d06e4ffd42ac6765baafb597a3c6bc5b53c3/0f1128a970bcba32416f4d78b7d1f9f20c585a98.patch) (3061 bytes) - added by adam.lofts on 2007-08-19 16:17:41Z - Patch to check for two null bytes after the header
* [SOCKSv4a_support.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/66d/66d5d06e4ffd42ac6765baafb597a3c6bc5b53c3/cc12a9bed257aedf5815db0adaaa5fd45f12a890.patch) (5860 bytes) - added by jesstess on 2009-09-07 02:51:01Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2791 2791
type__enhancement enhancement
reporter__adam_lofts adam.lofts
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__core core
keywords__ 
time__1187540225000000 1187540225000000
changetime__1265392999000000 1265392999000000
version__None None
owner__ 
cc__jesstess cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to adam.lofts|
|-|-|

Nice to know somebody is using that code!

If you'd like to speed this along, please read http://twistedmatrix.com/trac/wiki/TwistedDevelopment.  To start with, can you write some unit tests for the behavior you are changing so this bug doesn't resurface?

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess removed owner|
|-|-|

This really amounts to a feature request to support SOCKSv4A, I think.

The new patch fixes up a small initialization problem in adam.lofts' code as well as some cosmetic issues and adds a unit test in twisted/test/test_socks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

```twisted/test/test_socks.py``` needs a copyright header.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess removed owner|
|-|-|

So it does. Patch updated accordingly.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @jesstess|
|-|-|

Humm we had some overlap here.  [#3886](https://github.com/twisted/twisted/issues/3886) and [#3920](https://github.com/twisted/twisted/issues/3920) are pretty strongly related.  At least one of those may be a duplicate of this.  I reviewed this code anyway (I forgot about those tickets :), but it probably makes sense to take a look at the branch and patch associated with those tickets to make sure that continuing work here makes the most sense (all the redundant effort may have happened already, or there might be stuff worth taking from those tickets).

  1. On the new line 74 of twisted/protocols/socks.py, the `index` call can raise a `ValueError`.  This case isn't covered by the unit tests, though; it should be.
  1. The variables `idx1` and `idx2` in `SOCKSv4Incoming.dataReceived` are not very descriptively named. :)  These seem to be the ending positions of the user id and domain fields.
  1. Can you replace the `string.split` usage in this code with the `split` `str` method?
  1. In the unit tests, please use `assertEquals` rather than `assertEqual` (hm, is that in the coding standard yet?  I don't see it.)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set status to `closed`|
|-|-|

Fixed in [#3886](https://github.com/twisted/twisted/issues/3886).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
